### ISSUE TITLE
Cannot use single DataProvider for form and grid

### ISSUE THREAD

**Issue by gerti2112** [magento/magento2#5353]
## Steps to reproduce
1. Install Magento from `develop` branch.
2. Create custom module with custom DataProvider

``` php
class RegionDataProvider extends \Magento\Framework\View\Element\UiComponent\DataProvider\DataProvider
{
    ...
    /**
     * @return array
     */
    public function getMeta()
    {
        $meta = parent::getMeta();

        $meta['general'] = [
            'arguments' => [
                'data' => [
                    'config' => [
                        'componentType' => Fieldset::NAME,
                        'label' => __('FieldSet Example'),
                    ],
                ],
            ],
            'children' => [
                'title' => [
                    'arguments' => [
                        'data' => [
                            'config' => [
                                'label' => __('Field Example'),
                                'componentType' => Field::NAME,
                                'formElement' => 'input',
                            ],
                        ],
                    ],
                ],
            ],
        ];
        return $meta;
    }
}
```
1. Register for the form and for the grid SAME DataProvider in ui configuration

``` xml
<argument name="dataProvider" xsi:type="configurableObject">
            <argument name="class" xsi:type="string">MyApp\Region\Ui\DataProvider\RegionDataProvider</argument>
```
## Expected result
1. Field has been added to form
2. Grid is work
## Actual result
1. Field has been added to form properly
2. On grid page we have
   Fatal error: Call to a member function getComponentName() on null in .../app/code/Magento/Ui/Component/Form/Field.php on line 60

**Comment by gerti2112** [comment]
Quick steps to reproduce:
Open \Magento\Cms\Ui\Component\DataProvider and put next code

``` php
<?php
/**
 * Copyright © 2016 Magento. All rights reserved.
 * See COPYING.txt for license details.
 */
namespace Magento\Cms\Ui\Component;

use Magento\Framework\Api\FilterBuilder;
use Magento\Framework\Api\Search\SearchCriteriaBuilder;
use Magento\Framework\App\RequestInterface;
use Magento\Framework\View\Element\UiComponent\DataProvider\Reporting;

class DataProvider extends \Magento\Framework\View\Element\UiComponent\DataProvider\DataProvider
{
    /**
     * @param string $name
     * @param string $primaryFieldName
     * @param string $requestFieldName
     * @param Reporting $reporting
     * @param SearchCriteriaBuilder $searchCriteriaBuilder
     * @param RequestInterface $request
     * @param FilterBuilder $filterBuilder
     * @param array $meta
     * @param array $data
     */
    public function __construct(
        $name,
        $primaryFieldName,
        $requestFieldName,
        Reporting $reporting,
        SearchCriteriaBuilder $searchCriteriaBuilder,
        RequestInterface $request,
        FilterBuilder $filterBuilder,
        array $meta = [],
        array $data = []
    ) {
        parent::__construct(
            $name,
            $primaryFieldName,
            $requestFieldName,
            $reporting,
            $searchCriteriaBuilder,
            $request,
            $filterBuilder,
            $meta,
            $data
        );
    }

    /**
     * @return array
     */
    public function getMeta()
    {
        $meta = parent::getMeta();

        $meta['general'] = [
            'arguments' => [
                'data' => [
                    'config' => [
                        'componentType' => \Magento\Ui\Component\Form\Fieldset::NAME,
                        'label' => __('FieldSet Example'),
                    ],
                ],
            ],
            'children' => [
                'title' => [
                    'arguments' => [
                        'data' => [
                            'config' => [
                                'label' => __('Field Example'),
                                'componentType' => \Magento\Ui\Component\Form\Field::NAME,
                                'formElement' => 'input',
                            ],
                        ],
                    ],
                ],
            ],
        ];
        return $meta;
    }
}
```

And result we can see that cms page grid is broken

**Comment by magento-engcom-team** [comment]
@gerti2112, thank you for your report.
The issue is already fixed in develop branch, 2.2.0
